static_frame/core/index_hierarchy.py

```diff
@@ -220,122 +467,108 @@ def from_labels(cls: tp.Type[IH],
 
         Args:
             labels: an iterator or generator of tuples.
             *,
             name:
-            reorder_for_hierarchy: reorder the labels to produce a hierarchable Index, assuming hierarchability is possible.
+            reorder_for_hierarchy: an optional argument that will ensure the resulting index is arranged in a tree-like structure.
             index_constructors:
             depth_reference:
             continuation_token: a Hashable that will be used as a token to identify when a value in a label should use the previously encountered value at the same depth.
 
         Returns:
             :obj:`static_frame.IndexHierarchy`
         '''
-        if reorder_for_hierarchy:
-            if continuation_token != CONTINUATION_TOKEN_INACTIVE:
-                raise RuntimeError('continuation_token not supported when reorder_for_hiearchy')
-            # use from_records to ensure approprate columnar types
-            from static_frame import Frame
-            index_labels = Frame.from_records(labels)._blocks
-            # this will reorder and create the index using this smae method, passed as cls.from_labels
-            index, _ = rehierarch_from_type_blocks(
-                    labels=index_labels,
-                    depth_map=range(index_labels.shape[1]), # keep order
-                    index_cls=cls,
-                    index_constructors=index_constructors,
-                    name=name,
-                    )
-            return index #type: ignore
-
         labels_iter = iter(labels)
+
         try:
-            first = next(labels_iter)
-            labels_empty = False
+            label_row = next(labels_iter)
         except StopIteration:
-            labels_empty = True
-
-        if labels_empty:
-            # if iterable is empty, must discover depth
-            if isinstance(labels, np.ndarray) and labels.ndim == 2:
-                # if this is a 2D array, we can get the depth
-                depth = labels.shape[1]
-                if depth == 0: # an empty 2D array can have 0 depth
-                    pass # do not set depth_reference, assume it is set
-                elif ((depth_reference is None and depth > 1)
-                        or (depth_reference is not None and depth_reference == depth)):
-                    depth_reference = depth
+            labels_are_empty = True
+        else:
+            labels_are_empty = False
+
+        if labels_are_empty:
+            return cls._from_empty(labels, name=name, depth_reference=depth_reference)
+
+        depth = len(label_row)
+
+        if depth == 1:
+            raise ErrorInitIndex('Cannot create IndexHierarchy from only one level.')
+
+        # A mapping for each depth level, of label to index
+        hash_maps: tp.List[tp.Dict[tp.Hashable, int]] = [{} for _ in range(depth)]
+        indexers: tp.List[tp.List[int]] = [[] for _ in range(depth)]
+
+        prev_row: tp.Sequence[tp.Hashable] = EMPTY_TUPLE
```

Comment:
I lean towards no, since these types are only applicable to this specific context. I think it would harm readability to put these anywhere else.

Reply:
My suggestion here refers to my assumption above in regard to `tp` instance creation.

Reply:
I see. Do you think the loss of readability is worth it, given that this is a nanosecond cost happening in an otherwise expensive constructor? The construction of typehints is effectively 0 when compared to the cost of `from_labels`.

Reply:
I think the readability is improved!
```python
HasableToIntMaps = tp.List[tp.Dict[tp.Hashable, int]]
```